### ISSUE TITLE
Planner generates redundant redistribute motion for left outer join on distribution key

### ISSUE THREAD

**Issue by hsyuan** [greenplum-db/gpdb#4260]
Repro:
```sql
create table r0(a int, b int);
create table r1(a int, b int);
create table r2(a int, b int);
insert into r0 select i%100, i%100 from generate_series(1,10000) i;
insert into r1 select i%100, i%100 from generate_series(1,1000) i;
insert into r2 select i%100, i%100 from generate_series(1,100) i;
explain select * from r0 left join r1 on r0.a=r1.a left join r2 on r1.a=r2.a;

                                               QUERY PLAN
--------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=64.50..1326.50 rows=100000 width=24)
   ->  Hash Left Join  (cost=64.50..1326.50 rows=33334 width=24)
         Hash Cond: t0.a = t1.a
         ->  Seq Scan on t0  (cost=0.00..112.00 rows=3334 width=8)
         ->  Hash  (cost=52.00..52.00 rows=334 width=16)
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=5.25..52.00 rows=334 width=16)
                     Hash Key: t1.a
                     ->  Hash Left Join  (cost=5.25..32.00 rows=334 width=16)
                           Hash Cond: t1.a = t2.a
                           ->  Seq Scan on t1  (cost=0.00..13.00 rows=334 width=8)
                           ->  Hash  (cost=4.00..4.00 rows=34 width=8)
                                 ->  Seq Scan on t2  (cost=0.00..4.00 rows=34 width=8)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(14 rows)
```

Note that inner join doesn't generate redundant motion:
```sql
explain select * from t0 join t1 on t0.a=t1.a join t2 on t1.a=t2.a;
                                      QUERY PLAN
--------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=44.50..1306.50 rows=100000 width=24)
   ->  Hash Join  (cost=44.50..1306.50 rows=33334 width=24)
         Hash Cond: t0.a = t1.a
         ->  Seq Scan on t0  (cost=0.00..112.00 rows=3334 width=8)
         ->  Hash  (cost=32.00..32.00 rows=334 width=16)
               ->  Hash Join  (cost=5.25..32.00 rows=334 width=16)
                     Hash Cond: t1.a = t2.a
                     ->  Seq Scan on t1  (cost=0.00..13.00 rows=334 width=8)
                     ->  Hash  (cost=4.00..4.00 rows=34 width=8)
                           ->  Seq Scan on t2  (cost=0.00..4.00 rows=34 width=8)
 Settings:  optimizer=off
 Optimizer status: legacy query optimizer
(12 rows)
```

It looks like `cdbpath_match_preds_to_partkey_tail` generates the motion path for the inner LOJ.

**Comment by yydzero** [comment]
Still reproducible on latest master:

```
yydzero=# explain select * from r0 left join r1 on r0.a=r1.a left join r2 on r1.a=r2.a;
                                               QUERY PLAN
--------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  (cost=64.50..1326.50 rows=100000 width=24)
   ->  Hash Left Join  (cost=64.50..1326.50 rows=33334 width=24)
         Hash Cond: r0.a = r1.a
         ->  Seq Scan on r0  (cost=0.00..112.00 rows=3334 width=8)
         ->  Hash  (cost=52.00..52.00 rows=334 width=16)
               ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=5.25..52.00 rows=334 width=16)
                     Hash Key: r1.a
                     ->  Hash Left Join  (cost=5.25..32.00 rows=334 width=16)
                           Hash Cond: r1.a = r2.a
                           ->  Seq Scan on r1  (cost=0.00..13.00 rows=334 width=8)
                           ->  Hash  (cost=4.00..4.00 rows=34 width=8)
                                 ->  Seq Scan on r2  (cost=0.00..4.00 rows=34 width=8)
 Optimizer: legacy query optimizer
(13 rows)
```

**Comment by hlinnaka** [comment]
This was fixed with commit 3d6c78c90a6eac69aa8a82f7a51ecd1183a3c771.
